src/bundles/ipfs-provider.js

```diff
@@ -0,0 +1,271 @@
+    try {
+      setting = window.localStorage.getItem(id)
+    } catch (error) {
+      console.error(`Error reading '${id}' value from localStorage`, error)
+    }
+
+    try {
+      return JSON.parse(setting)
+    } catch (_) {
+      // res was probably a string, so pass it on.
+      return setting
+    }
+  }
+}
+
+const writeSetting = (id, value) => {
+  try {
+    window.localStorage.setItem(id, JSON.stringify(value))
+  } catch (error) {
+    console.error(`Error writing '${id}' value to localStorage`, error)
+  }
+}
+
+/**
+ * @typedef {Object} IPFSAPI
+ * @property {(callback?:Function) => Promise<void>} stop
+ */
+
+/** @type {IPFSAPI|void} */
+let ipfs = null
+
+/**
+ * @typedef {Object} State
+ * @property {Model} ipfs
+ */
+
+const bundle = {
+  name: 'ipfs',
+  reducer: (state, message) => update(state == null ? init() : state, message),
+  getExtraArgs () {
+    return { getIpfs: () => ipfs }
+  },
+  /**
+   * @param {State} state
+   */
+  selectIpfsReady: state => state.ipfs.ready,
+  /**
+   * @param {State} state
+   */
+  selectIpfsProvider: state => state.ipfs.provider,
+  /**
+   * @param {State} state
+   */
+  selectIpfsApiAddress: state => state.ipfs.apiAddress,
+  /**
+   * @param {State} state
+   */
+  selectIpfsInvalidAddress: state => state.ipfs.invalidAddress,
+  /**
+   * @param {State} state
+   */
+  selectIpfsInitFailed: state => state.ipfs.failed,
+
+  doInitIpfs: () => async (store) => {
+    await initIPFS(store)
+  },
+
+  doStopIpfs: () => async (store) => {
+    if (ipfs) {
+      ipfs.stop(() => {
+        store.dispatch({ type: 'IPFS_STOPPED' })
+      })
+    }
+  },
+
+  doUpdateIpfsApiAddress: (address) => async (store) => {
+    const apiAddress = asAPIAddress(address)
+    if (apiAddress == null) {
+      store.dispatch({ type: 'IPFS_API_ADDRESS_INVALID' })
+    } else {
+      await writeSetting('ipfsApi', apiAddress)
+      store.dispatch({ type: 'IPFS_API_ADDRESS_UPDATED', payload: apiAddress })
+
+      await initIPFS(store)
+    }
+  },
+
+  doDismissIpfsInvalidAddress: () => (store) => {
+    store.dispatch({ type: 'IPFS_API_ADDRESS_INVALID_DISMISS' })
+  }
+}
+
+const initIPFS = async (store) => {
+  store.dispatch({ type: 'IPFS_INIT_STARTED' })
+
+  /** @type {Model} */
+  const { apiAddress } = store.getState().ipfs
+
+  try {
+    const result = await getIpfs({
+      // @ts-ignore - TS can't seem to infer connectionTest option
+      connectionTest: async (ipfs) => {
+        // ipfs connection is working if can we fetch the bw stats.
+        // See: https://github.com/ipfs-shipyard/ipfs-webui/issues/835#issuecomment-466966884
+        try {
+          await ipfs.stats.bw()
+        } catch (err) {
+          if (!/bandwidth reporter disabled in config/.test(err)) {
+            throw err
+          }
+        }
+
+        return true
+      },
+      loadHttpClientModule: () => HttpClient,
+      providers: [
+        providers.httpClient({ apiAddress })
```

Comment:
I was under impression that ipfs-provider did that already:
https://github.com/ipfs-shipyard/ipfs-provider/blob/3e275b3aa79788a3adce1e3fb3c0b1fbfff0d972/src/providers/http-client.js#L43-L53

As an aside it would be nice if `ipfs-provider` had an API that was easier to introspect, e.g.

```js
const provider = new IPFSProvider({
    loadHttpClientModule: () => window.IpfsHttpClient,
    loadJsIpfsModule: () => window.Ipfs,
    providers: [ /* see Usage below */ ]
})

console.log(provider.strategy) // => prints out ever strategy that `provide()` will try until one succeeds.

const { ipfs } = await provider.provide()
```

Reply:
Ah, right, the fallback to `origin` kicks in  if `apiAddress` is undefined:

```js
// Explicit custom apiAddress provided. Only try that.
  if (apiAddress) {
    return maybeApi({ apiAddress, connectionTest, httpClient })
  }
  // fallback to origin
```
 
In that case disregard  my comment :) 

(strategy inspector for ipfs-provider is a good idea – feel free to PR, you should have write access to the repo too)